### ISSUE TITLE
trouble with windows install (update for alexa-app-server/alexa-app v3.0.0)

### ISSUE THREAD

**Issue by greenpad2** [jradwan/alexa_tivo_control#20]
While I use linux at work, I have windows at home, so I'm attempting to install on windows 10.  I'm getting the following error when I try to run.  I've verified that express is installed and I've tried several versions of nvm.  Thanks for any help! 

```
D:\Program Files\nodejs\node_modules\alexa-app-server\examples>node server
serving static content from: public_html
loading server-side modules from: server
   loaded D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\examples\server\login.js
loading apps from: apps
[2017-02-09T22:31:15.724Z] Currently controlling: Living Room (192.168.1.172)
D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\examples\apps\alexa_tivo_control\node_modules\alexa-app\index.js:491
    express.post(endpoint, function(req, res) {
            ^

TypeError: express.post is not a function
    at Object.express (D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\examples\apps\alexa_tivo_control\node_modules\alexa-app\index.js:491:13)
    at D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\index.js:116:11
    at Array.forEach (native)
    at Object.self.load_apps (D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\index.js:76:30)
    at Object.self.start (D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\index.js:186:12)
    at Function.appServer.start (D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\index.js:294:21)
    at Object.<anonymous> (D:\Users\keith\AppData\Roaming\nvm\v7.5.0\node_modules\alexa-app-server\examples\server.js:2:16)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:418:7)
    at startup (bootstrap_node.js:139:9)
    at bootstrap_node.js:533:3
```

**Comment by jradwan** [comment]
I did all my testing on Windows (and based the installation instructions) using nvm 0.10.36. I see you've tried 0.10.38 and above it looks like 7.5.0? 

Can you try 0.10.36 and see if that works? Also, what port are you using in server.js? Maybe try changing that (I use 8196, for example).

**Comment by greenpad2** [comment]
Thanks for the response.  I went back and tried 0.10.36 again with the same result.  I copied my server.js from your instructions page, except I put in ../index.js because I assume ./index.js is a typo.  I also tried 8080 to no effect.  Any other suggestions?  Thanks!

**Comment by jradwan** [comment]
You're right about the typo ... I've fixed that, thanks.

And you're sure express is installed? If you do an 'npm list' from D:\Users\keith\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\ you should see express somewhere in the tree:

```
├─┬ express@4.0.0
│ ├─┬ accepts@1.0.0
│ │ ├── mime@1.2.11
│ │ └── negotiator@0.3.0
```

That's all I can come up with since the error from index.js:

```
express.post(endpoint, function(req, res) {
            ^
TypeError: express.post is not a function
```

Seems to indicate a problem with calling express.post. Although index.js does include:

```
var express = require('express');
```

So I would think if it couldn't find express it would error out earlier.

**Comment by greenpad2** [comment]
Yes:

```
├─┬ express@4.14.1
│ ├─┬ accepts@1.3.3
│ │ ├─┬ mime-types@2.1.14
│ │ │ └── mime-db@1.26.0
│ │ └── negotiator@0.6.1
│ ├── array-flatten@1.1.1
```
Any way I can test to see if the port is the problem?

Thanks again for your help

**Comment by jradwan** [comment]
Hmmm ... looking at index.js I'm thinking it may not be the port but your "endpoint" which index.js defines as:

```
var endpoint = (root||'/') + (app.endpoint || app.name);
```

Try adding this line to index.js right before the express.post call (line 95):

```
self.log("*** endpoint = " + endpoint);
```

You should see then see a line like this in the console log after starting node-js:

```
[2017-02-10T20:17:27.925Z] *** endpoint = /tivo_control
```

I'm wondering if maybe your endpoint is invalid.

**Comment by greenpad2** [comment]
Thanks.  I don't actually see that line in my code--I have on line 490 of alexa-tivo-control/node_modules/alexa_app/index.js:
`var endpoint = (path || "/") + (self.endpoint || self.name);`
I don't see anything else close to what you've listed so I tried the log there:
```
        self.log("*** endpoint = " + endpoint);
             ^
TypeError: Object [object Object] has no method 'log'
```
Is there a global object I can use for logging?  Sorry you're having to walk me through basic js.  Thanks

**Comment by jradwan** [comment]
Whoops ... wrong index.js. I was looking at the one for alexa-app-server and not under alexa-app. 

Try this on line 489 (before the this.express line):

```
console.log("*** endpoint: /" + (self.endpoint || self.name));
```

That should give you something like this:

```
Loading apps from: apps
[2017-02-10T21:55:02.658Z] Currently controlling: Office (192.168.1.73)
[2017-02-10T21:55:02.659Z] *** endpoint: /tivo_control
[2017-02-10T21:55:02.660Z]    Loaded app [alexa_tivo_control] at endpoint: /tivo_control
[2017-02-10T21:55:02.667Z] Listening on HTTP port 8196
```

I'm no js expert either, so I'm feeling this all out as I go along. 😉 You should be able to pepper in some console.log lines to print out more debugging messages.

**Comment by greenpad2** [comment]
It looks like there is a valid endpoint:
`[2017-02-10T22:15:18.110Z] Currently controlling: Living Room (192.168.1.172)
[2017-02-10T22:15:18.112Z] *** endpoint: /tivo_control`
I did try deleting the tivo app directory and with the default apps, it connected to the port without an error, so I don't think it's the port.

**Comment by greenpad2** [comment]
Small progress.  I added "var app=express();" and got:
```
   loaded D:\Users\keith\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\server\login.js
loading apps from: apps
[2017-02-10T22:40:09.280Z] Currently controlling: Living Room (192.168.1.172)
[2017-02-10T22:40:09.281Z]    error loading app [D:\Users\keith\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\apps\alexa_tivo_control\app.js]: ReferenceError: express is not defined
[2017-02-10T22:40:09.349Z]    loaded app [hello_world] at endpoint: /alexa/hello_world
```
So it looks like it's not loading express, even though I have:
`var express = require('express');`
in alexa-app-server/index.js

**Comment by jradwan** [comment]
Strange. Unless it's another version thing? I seem to have 4.0.0 and you've got 4.14.1. That's probably not it. I'm not sure what the issue is here. I've done installs on Linux and Windows (when writing the installation instructions) and didn't have this particular problem. Hmm ...

**Comment by jradwan** [comment]
@spherular has it working under Windows (see issue #21). Maybe you can two can compare notes. Sorry I haven't been able to figure this one out for you ...

**Comment by spherular** [comment]
Yes mine is working under windows. Mine ran out of the box, just getting it live on Amazon was my sticking point.

**Comment by greenpad2** [comment]
Thanks to you both. I am actually traveling for the next week so I won't be able to look at it. I will try again when I get back.

**Comment by Dillonmartine** [comment]
Just to add input, I am getting a very similar error on both Windows and Linux. 

c:\Program Files\nodejs\node_modules\alexa-app-server\examples>node server.js
serving static content from: public_html
loading server-side modules from: server
   loaded C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\node_modules\alexa
-app-server\examples\server\login.js
loading apps from: apps
[2017-02-13T16:58:19.592Z] Currently controlling: Martine57 (192.168.1.105)

C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\node_modules\alexa-app-serve
r\examples\apps\alexa_tivo_control\node_modules\alexa-app\index.js:491
    express.post(endpoint, function(req, res) {
            ^
TypeError: Object #<Object> has no method 'post'
    at Object.express (C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\node_
modules\alexa-app-server\examples\apps\alexa_tivo_control\node_modules\alexa-app
\index.js:491:13)
    at C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\node_modules\alexa-ap
p-server\index.js:116:11
    at Array.forEach (native)
    at Object.self.load_apps (C:\Users\Administrator\AppData\Roaming\nvm\v0.10.4
0\node_modules\alexa-app-server\index.js:76:30)
    at Object.self.start (C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\no
de_modules\alexa-app-server\index.js:186:12)
    at Function.appServer.start (C:\Users\Administrator\AppData\Roaming\nvm\v0.1
0.40\node_modules\alexa-app-server\index.js:294:21)
    at Object.<anonymous> (C:\Users\Administrator\AppData\Roaming\nvm\v0.10.40\n
ode_modules\alexa-app-server\examples\server.js:2:16)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:935:3

c:\Program Files\nodejs\node_modules\alexa-app-server\examples>

I've tried various versions with no luck

**Comment by jradwan** [comment]
Ok, I did a clean install on a different Windows PC following my wiki instructions and am seeing the same error. I'll have to play around and see if I can figure out what's causing this ...

```
C:\Users\jeremy>nvm install v0.10.36
Downloading node.js version 0.10.36 (64-bit)... Complete
Creating C:\Users\jeremy\AppData\Roaming\nvm\temp

Downloading npm version 1.4.28... Complete
Installing npm v1.4.28...

Installation complete. If you want to use this version, type

nvm use 0.10.36

C:\Users\jeremy>nvm use 0.10.36
Now using node v0.10.36 (64-bit)

C:\Users\jeremy>cd \progra~1\nodejs

C:\PROGRA~1\nodejs>npm install alexa-app-server
npm WARN engine alexa-app@3.1.0: wanted: {"node":">=0.12"} (current: {"node":"0.10.36","npm":"1.4.28"})
npm WARN engine request@2.79.0: wanted: {"node":">= 4"} (current: {"node":"0.10.36","npm":"1.4.28"})
npm WARN engine form-data@2.1.2: wanted: {"node":">= 0.12"} (current: {"node":"0.10.36","npm":"1.4.28"})
npm WARN engine hoek@2.16.3: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.36","npm":"1.4.28"})
npm WARN engine cryptiles@2.0.5: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.36","npm":"1.4.28"})
npm WARN engine boom@2.10.1: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.36","npm":"1.4.28"})
alexa-app-server@3.0.0 node_modules\alexa-app-server
├── lodash.defaults@4.2.0
├── ejs@2.5.5
├── bluebird@3.4.7
├── body-parser@1.16.1 (content-type@1.0.2, bytes@2.4.0, depd@1.1.0, qs@6.2.1, raw-body@2.2.0, on-finished@2.3.0, http-errors@1.5.1, iconv-lite@0.4.15, debug@2.6.1, type-is@1.6.14)
├── alexa-app@3.1.0 (numbered@1.0.0, alexa-utterances@0.2.1, bluebird@2.11.0)
├── express@4.14.1 (escape-html@1.0.3, array-flatten@1.1.1, utils-merge@1.0.0, fresh@0.3.0, cookie-signature@1.0.6, methods@1.1.2, content-type@1.0.2, merge-descriptors@1.0.1, path-to-regexp@0.1.7, vary@1.1.0, range-parser@1.2.0, parseurl@1.3.1, encodeurl@1.0.1, etag@1.7.0, cookie@0.3.1, content-disposition@0.5.2, serve-static@1.11.2, depd@1.1.0, qs@6.2.0, on-finished@2.3.0, finalhandler@0.5.1, debug@2.2.0, send@0.14.2, proxy-addr@1.1.3, type-is@1.6.14, accepts@1.3.3)
├── alexa-verifier-middleware@0.1.9 (alexa-verifier@0.3.0)
└── hotswap@1.1.0 (async@2.1.4)
```

Then, after extracting alexa-tivo-control:

```
C:\PROGRA~1\nodejs\node_modules\alexa-app-server\examples\apps\alexa_tivo_control>npm install
ejs@2.5.5 node_modules\ejs

log-timestamp@0.1.2 node_modules\log-timestamp
└── log-prefix@0.0.0

telnet-client@0.9.4 node_modules\telnet-client
└── bluebird@3.0.5

alexa-app@2.4.0 node_modules\alexa-app
├── numbered@1.0.0
├── alexa-utterances@0.2.1 (js-combinatorics@0.5.2)
└── bluebird@2.11.0

C:\PROGRA~1\nodejs\node_modules\alexa-app-server\examples>node server.js
serving static content from: public_html
loading server-side modules from: server
   loaded C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\server\login.js
loading apps from: apps
[2017-02-13T23:42:26.655Z] Currently controlling: Office (192.168.1.73)

C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\apps\alexa_tivo_control\node_modules\alexa-app\index.js:491
    express.post(endpoint, function(req, res) {
            ^
TypeError: Object #<Object> has no method 'post'
    at Object.express (C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\apps\alexa_tivo_control\node_modules\alexa-app\index.js:491:13)
    at C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\index.js:116:11
    at Array.forEach (native)
    at Object.self.load_apps (C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\index.js:76:30)
    at Object.self.start (C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\index.js:186:12)
    at Function.appServer.start (C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\index.js:294:21)
    at Object.<anonymous> (C:\Users\jeremy\AppData\Roaming\nvm\v0.10.36\node_modules\alexa-app-server\examples\server.js:3:16)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:929:3
```

**Comment by jradwan** [comment]
Ok, I think I've found the problem. The current version of alexa-app-server is [3.0.0](https://github.com/alexa-js/alexa-app-server/blob/master/UPGRADING.md), which requires a minimum alexa-app version of 3.0.0 (or currently [3.1.0](https://github.com/alexa-js/alexa-app/blob/master/CHANGELOG.md)).

They made [changes](https://github.com/alexa-js/alexa-app/blob/master/UPGRADING.md) to the Express stuff in alexa-app 3.0.0 and that's what I believe is causing this problem.

For now, the workaround is to use alexa-app-server [2.3.1](https://github.com/alexa-js/alexa-app-server/releases/tag/v2.3.1). which will in turn use alexa-app [2.4.0](https://github.com/alexa-js/alexa-app/releases/tag/v2.4.0) which will work with my current alexa-tivo-control codebase. Whew! 

I'll see if I can make my code compatible with the latest alexa-app-server and alexa-app releases. They seem to be changing faster than I can keep up. 😮

**Comment by jradwan** [comment]
So actually this problem is due to a mismatch between the required version of alexa-app for alexa-app-server. If you follow my installation instructions, you'll get the current version of alexa-app-server (3.0.0). Then, when you 'npm install' to get all the dependencies for alexa_tivo_control, the package.json file says:

```
  "dependencies": {
    "alexa-app": "^2.4.0",
```

which downloads alexa-app 2.4.0. But alexa-app-server 3.0.0 requires at a minimum alexa-app 3.0.0 (currently 3.1.0). Running alexa-app 2.4.0 with alexa-app-server 3.0.0 causes the express error.

So, if you use current versions of alexa-app-server and alexa-app, it should be fine. I'll updated package.json to reflect this as well. No other code changes should be necessary ... but I'll do some more testing.

**Comment by jradwan** [comment]
I've done fresh installs on Linux and Windows using alexa-app-server v3.0.0 and alexa-app v3.1.0 and everything seems to be fine. This also requires using a newer version of nodejs (not 0.10.36 that I've been using up until now). So I've used the current v6.9.5 release and it seems to work so I'm sticking with that.

I've updated the installation instructions accordingly. Sorry for all the confusion! It took me a while to parse through all the changes they've made to alexa-app-server and alexa-app recently.
